OpenChart/src/Formats/OpenChart/v0.1/OpenChartConverter.cs

```diff
@@ -70,11 +120,79 @@ public ProjectData FromNative(Project project)
             {
                 var chart = new ChartData();
+                chartList.Add(chart);
+
                 chart.KeyCount = c.KeyCount;
+                chart.Author = c.Author;
+                chart.ChartName = c.ChartName;
+                chart.BPMs = c.BPMs.ToArray();
+
+                // Beat rows are stored as a dictionary. This is necessary because the
+                // native Chart class stores objects by key index, whereas the file
+                // format stores them by row. As we iterate through each key index we
+                // need to be able to retrieve the beat row at that particular beat.
+                var dict = new Dictionary<Beat, BeatRowData>();
+
+                // Iterate through each key index.
+                for (var keyIndex = 0; keyIndex < chart.KeyCount.Value; keyIndex++)
+                {
+                    // Get each object for this key index.
+                    foreach (var keyObject in c.Objects[keyIndex])
+                    {
+                        BeatRowData row;
+                        FormatObjects.IChartObject obj;
+
+                        // Try getting the beat row, or create one if it doesn't exist yet.
+                        if (!dict.ContainsKey(keyObject.Beat))
+                        {
+                            row = new BeatRowData(chart.KeyCount.Value);
+                            row.Beat = keyObject.Beat;
+                            dict[keyObject.Beat] = row;
+                        }
+                        else
+                            row = dict[keyObject.Beat];
+
+                        // Map to the appropriate object type.
+                        if (keyObject is NativeObjects.TapNote)
+                            obj = new FormatObjects.TapNote();
+                        else if (keyObject is NativeObjects.HoldNote holdNote)
+                        {
+                            obj = new FormatObjects.HoldNote
+                            {
+                                Length = holdNote.Length
+                            };
+                        }
+                        else
+                            throw new ConverterException("Unknown object type");
+
+                        row.Objects[keyIndex] = obj;
+                    }
+                }
+
+                // Get the beat rows from the dictionary and sort them by beat.
+                var rowList = new List<BeatRowData>(dict.Count);
+
+                rowList.AddRange(dict.Values);
+                rowList.Sort();
+
+                chart.Rows = rowList.ToArray();
             }
 
             pd.Charts = chartList.ToArray();
 
+            try
+            {
+                pd.Validate();
+            }
+            catch (Exception e)
+            {
+                throw new ConverterException(
+                    "An error occurred while preparing to write a project to a file. " +
+                    "(This shouldn't happen!)",
```

Comment:
This feels like "captain obvious" error logging - just because it _shouldn't_ happen doesn't necessarily mean it _can't_

Reply:
I wanted to call out the fact that this is purely an internal error, not something the user did wrong. Whatever is causing that to throw should have been caught by something else upstream